File: packages/dll/src/net/messages/getinstances.cpp
#include "getinstances.h"
#include "../message.h"
#include "../../injector.h"
#include "../../UE/v425/uobjectarray.h"
#include "../../UE/v425/uobject.h"

GetInstancesRequest::GetInstancesRequest(Buffer& data) : Request(data), name(data.Read()), count(data.Read<uint32_t>()), deep(data.Read<bool>()) {}

void GetInstancesRequest::Process() {
	GetInstancesResponse res;

	WFUObjectItem item = injector.objectArray.FindObject(name);

	if (item) {
		std::vector<WFUObjectItem> instances = injector.objectArray.FindInstances(item.GetObject(), count, deep);
		for (auto instance : instances) {
			if (!instance)
				continue;

			res.names.push_back(instance.GetObject().GetFullName());
		}
	}

	ProcessResponse(res);
}

void GetInstancesResponse::Process(Buffer& data) {
	Response::Process(data);

	data.Write(names.size());
	for (auto& name : names)
	{
		data.Write(name);
	}
}

File: packages/electron/src/main/net/messages/getInstances.ts
import { RROxApp } from "../../app";
import { BufferIO } from "../io";
import { MessageType, Request, Response } from "../message";

export class GetInstancesRequest extends Request {
    public name: string;
    public count: number;
    public deep: boolean;

    constructor( app: RROxApp, name: string, count: number = 0, deep: boolean = false ) {
        super( app, MessageType.GET_INSTANCES );

        this.name = name;
        this.count = count;
        this.deep = deep;
    }

    public process( data: BufferIO ) {
        super.process( data );

        data.writeString( this.name );
        data.writeUInt32( this.count );
        data.writeBool( this.deep );
    }
}

export class GetInstancesResponse extends Response {
    public list: string[]

    constructor( app: RROxApp, data: BufferIO ) {
        super( app, data );

        this.list = [];
        
        data.readArray( () => {
            const str = data.readString();
            if( str )
                this.list.push( str );
        } );
    }
}